### PR TITLE
Report nullability warnings for user-defined conversions in tuples

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -3756,51 +3761,16 @@ private void TrackNullableStateOfNullableValue(int containingSlot, TypeSymbol co
...
-            }
-
-            int slot = GetOrCreateObjectCreationPlaceholderSlot(node);
-            if (slot < 0)
-            {
-                return;
-            }
-
-            TrackNullableStateOfTupleConversion(node.Conversion, node.Type, node.Operand.Type, slot, valueSlot);
-        }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: These three methods were inlined in the one caller in each case.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -4038,76 +4018,7 @@ private bool HasTopLevelNullabilityConversion(TypeWithAnnotations source, TypeWi
...
-
-                        // conversion "to" type -> final type
-                        // https://github.com/dotnet/roslyn/issues/29959 If the original conversion was
-                        // explicit, this conversion should not report nested nullability mismatches.
-                        // (see NullableReferenceTypesTests.ExplicitCast_UserDefined_02).
-                        operandType = ClassifyAndApplyConversion(node, targetTypeWithNullability, operandType,
-                            useLegacyWarnings, assignmentKind, target, reportWarnings: reportRemainingWarnings);
-                        return operandType;
-                    }
+                    return ApplyUserDefinedConversion(node, operandOpt, conversion, targetTypeWithNullability, operandType, useLegacyWarnings, assignmentKind, target, reportTopLevelWarnings, reportRemainingWarnings);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Code extracted, unchanged, to `ApplyUserDefinedConversion` method.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -4299,6 +4230,99 @@ private bool HasTopLevelNullabilityConversion(TypeWithAnnotations source, TypeWi
...
+            bool reportRemainingWarnings)
+        {
+            Debug.Assert(!IsConditionalState);
+            Debug.Assert(node != null);
+            Debug.Assert(operandOpt != null || !operandType.HasNullType);
+            Debug.Assert(targetTypeWithNullability.HasType);
+            Debug.Assert((object)target != null || assignmentKind != AssignmentKind.Argument);
+
+            TypeSymbol targetType = targetTypeWithNullability.Type;
+            Debug.Assert(conversion.Kind == ConversionKind.ExplicitUserDefined || conversion.Kind == ConversionKind.ImplicitUserDefined);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: nit: group this with the other asserts at the top of the method?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Will move in next PR.

---
In reply to: [269783790](https://github.com/dotnet/roslyn/pull/34496#discussion_r269783790) [](ancestors = 269783790)

**Pull request merged**